ui/revenuecatui/src/test/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModelTest.kt

```diff
@@ -0,0 +1,267 @@
+            // Delay response to verify initial state
+            delay(100)
+            offerings
+        }
+
+        val model = create()
+
+        when (model.state.value) {
+            is PaywallState.Loading -> {}
+            is PaywallState.Error,
+            is PaywallState.Loaded,
+            -> fail("Invalid state")
+        }
+
+        assertThat(model.actionInProgress.value).isFalse
+    }
+
+    @Test
+    fun `Should load default offering`() {
+        val model = create(activeSubscriptions = arrayOf(TestData.Packages.monthly.product.id))
+
+        coVerify { purchases.awaitOfferings() }
+
+        val state = model.state.value
+        if (state !is PaywallState.Loaded) {
+            fail("Invalid state")
+            return
+        }
+
+        val expectedPaywall = defaultOffering.paywall!!
+
+        verifyPaywall(state, expectedPaywall)
+        assertThat(state.templateConfiguration.packages.packageIsCurrentlySubscribed(TestData.Packages.monthly))
+            .isTrue
+        assertThat(state.templateConfiguration.packages.packageIsCurrentlySubscribed(TestData.Packages.annual))
+            .isFalse
+    }
+
+    @Test
+    fun `Should load selected offering`() {
+        val offering = TestData.template1Offering
+        val model = create(offering = offering)
+
+        coVerify(exactly = 0) { purchases.awaitOfferings() }
+
+        val state = model.state.value
+        if (state !is PaywallState.Loaded) {
+            fail("Invalid state")
+            return
+        }
+
+        val expectedPaywall = offering.paywall!!
+
+        verifyPaywall(state, expectedPaywall)
+    }
+
+    @Test
+    fun `selectPackage`() {
+        val model = create()
+
+        val state = model.state.value
+        if (state !is PaywallState.Loaded) {
+            fail("Invalid state")
+            return
+        }
+
+        val packageToSelect = state.templateConfiguration.packages.all[0]
+        assertThat(packageToSelect).isNotEqualTo(state.selectedPackage.value)
+
+        model.selectPackage(packageToSelect)
+        assertThat(state.selectedPackage.value).isSameAs(packageToSelect)
+    }
+
+    @Test
+    fun `purchasePackage`() {
+        val model = create()
+
+        val state = model.state.value
+        if (state !is PaywallState.Loaded) {
+            fail("Invalid state")
+            return
+        }
+
+        val transaction = mockk<StoreTransaction>()
+        val selectedPackage = state.selectedPackage.value
+
+        coEvery {
+            purchases.awaitPurchase(any())
+        } returns PurchaseResult(transaction, customerInfo)
+
+        model.purchaseSelectedPackage(context)
+
+        coVerify {
+            purchases.awaitPurchase(any())
+        }
+
+        verifyOrder {
+            listener.onPurchaseStarted(selectedPackage.rcPackage)
+            listener.onPurchaseCompleted(customerInfo, transaction)
+        }
+
+        assertThat(model.actionInProgress.value).isFalse
```

Comment:
I guess we are not verifying the changes in this value, but I'm unsure on what's the best way to do this with coroutines... (with callbacks, I would hold a reference to the callback and call it at the appropriate time)... With coroutines, I can only think of delaying the response to verify this field, but not really a great option... Any ideas @vegaro? I don't think this is a blocker though.

Reply:
Yeah I tried different things but couldn't figure it out.
Definitely a useful addition to these tests though.